File: Data Structures & Algorithms/binary-tree-maximum-path-sum/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        res = [-float('inf')]

        def dfs(root):
            if not root:
                return 0
            
            leftmax = max(0, dfs(root.left))
            rightmax = max(0, dfs(root.right))

            #Split
            res[0] = max(res[0], root.val + leftmax + rightmax)
            return root.val + max(leftmax, rightmax)
        
        dfs(root)
        return res[0]
        

File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int goodNodes(TreeNode* root) {
        dfs(root, -999);
        return res;
    }
private:
    int res = 0;
    void dfs(TreeNode* node, int max){
        if (node != nullptr){
            if (node->val >= max){
                res += 1;
                max = node->val;
            }
            dfs(node->left, max);
            dfs(node->right, max);
        }
    }
};
